### PR TITLE
[SWDEV-321331] Fix ONNX runtime unit tests for reductions

### PR DESCRIPTION
**Pull request by qianfengz**
Quick fix for [SWDEV-321331](https://ontrack-internal.amd.com/browse/SWDEV-321331). 

The fix has been verified with the following script using MIOpenDriver. 
```bash
#!/bin/bash

./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 8,2,1 -I 0 -N 1 -O 0 -R 0,1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 160,10,1 -I 0 -N 1 -O 0 -R 0,1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 7,1024,1 -I 0 -N 1 -O 0 -R 0,1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 1024,30528,1 -I 0 -N 1 -O 0 -R 0,1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 3,1,1 -I 0 -N 1 -O 0 -R 0,1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 3,1,1 -I 0 -N 1 -O 1 -R 0,1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 3,1,1 -I 1 -N 1 -O 3 -R 0,1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 3,2,1 -I 1 -N 1 -O 3 -R 1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 6,2,1 -I 0 -N 1 -O 3 -R 1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 6,2,1 -I 0 -N 1 -O 2 -R 1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 2,2,1 -I 0 -N 1 -O 0 -R 1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 4,3,1 -I 0 -N 1 -O 3 -R 1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 3,4,1 -I 0 -N 1 -O 3 -R 1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 3,4,1 -I 0 -N 1 -O 3 -R 0,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 2048,32,1 -I 0 -N 1 -O 3 -R 0,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 4,3,1 -I 0 -N 1 -O 2 -R 1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 3,4,1 -I 0 -N 1 -O 2 -R 0,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 2048,32,1 -I 0 -N 1 -O 2 -R 0,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 3,4,1 -I 0 -N 1 -O 2 -R 0,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 12,11,1 -I 0 -N 1 -O 0 -R 0,1,2
./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 13,4,7,7 -I 0 -N 1 -O 0 -R 0,1,2,3

```

The usual `bin/test_reduce_test --all`
The fix has also been verified with a script using /bin/test_reduce_all: 
```bash
#!/bin/bash

./bin/test_reduce_test --scales 1 0 --CompType 1 --D 8 2 1 --I 0 --N 1 ---ReduceOp 0 --R 0 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 160 10 1 --I 0 --N 1 ---ReduceOp 0 --R 0 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 7 1024 1 --I 0 --N 1 ---ReduceOp 0 --R 0 1 2 -v
## The case will fail due to big accumulative error, so not used with test_reduce_test
##./bin/test_reduce_test --scales 1 0 --CompType 1 --D 1024 30528 1 --I 0 --N 1 ---ReduceOp 0 --R 0 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 3 1 1 --I 0 --N 1 ---ReduceOp 0 --R 0 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 3 1 1 --I 0 --N 1 ---ReduceOp 1 --R 0 1 2 -v 
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 3 1 1 --I 1 --N 1 ---ReduceOp 3 --R 0 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 3 2 1 --I 1 --N 1 ---ReduceOp 3 --R 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 6 2 1 --I 0 --N 1 ---ReduceOp 3 --R 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 6 2 1 --I 0 --N 1 ---ReduceOp 2 --R 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 2 2 1 --I 0 --N 1 ---ReduceOp 0 --R 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 4 3 1 --I 0 --N 1 ---ReduceOp 3 --R 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 3 4 1 --I 0 --N 1 ---ReduceOp 3 --R 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 3 4 1 --I 0 --N 1 ---ReduceOp 3 --R 0 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 2048 32 1 --I 0 --N 1 ---ReduceOp 3 --R 0 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 4 3 1 --I 0 --N 1 ---ReduceOp 2 --R 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 3 4 1 --I 0 --N 1 ---ReduceOp 2 --R 0 2 -v 
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 2048 32 1 --I 0 --N 1 ---ReduceOp 2 --R 0 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 3 4 1 --I 0 --N 1 ---ReduceOp 2 --R 0 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 12 11 1 --I 0 --N 1 ---ReduceOp 0 --R 0 1 2 -v
./bin/test_reduce_test --scales 1 0 --CompType 1 --D 13 4 7 7 --I 0 --N 1 ---ReduceOp 0 --R 0 1 2 3 -v
```

### CODE REVIEW

**Comment by qianfengz** [comment]
Explanation:

1. The issue fixed by Commit [2083](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1409/commits/bf72f62c739fd81f73d06c0720f8e0a2317f2083) will cause reduce_partial_dims kernel be called even though reduce_all_dims kernel is needed when the extent length of some dimension is 1
2. The issue fixed by Commit [4531](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1409/commits/773cee8e4cde9bc4ee5184b6a1c3ec0415494531) will cause kernel compile-time error since the useless type declaration could not be satisfied due to the absence of the depending compiler constants from the host.

**Comment by qianfengz** [comment]
@atamazov  should some verification tests be added to test/CMakeLists.txt ?   Can we using MIOpenDriver to construct a test command in test/CMakeLists.txt ?    I have not seen any case which uses MIOpenDriver in test/CMakeLists.txt

**Comment by junliume** [comment]
@atamazov there are some test verification errors (***verify_reduce_no_indices failed***):
```
./bin/test_reduce_test --float --D 8 2 1 --R 0 1 2 --ReduceOp 0 --CompType 1 --N 0 --I 0 --scales 1 0
error: 0
Max diff: 0
verify_reduce_no_indices failed
Input Tensor 8, 2, 1

```
at the same time the MIOpenDriver commands succeeded:
```
root@hpe-sjc2-43:/root/MIOpen/build# ./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 8,2,1 -I 0 -N 1 -O 0 -R 0,1,2
MIOpenDriver reduce -A 1 -B 0 -C 1 -D 8,2,1 -I 0 -N 1 -O 0 -R 0,1,2
ReduceTensor() Verifies on CPU and GPU (err=0.000000)

**Comment by atamazov** [comment]
@junliume I am looking into tests. Stay tuned.

**Comment by atamazov** [comment]
`verify_reduce_no_indices failed` is not an actual failure

**Comment by atamazov** [comment]
All test cases show progressions from "FAILED" to "error: 0"

**Review by atamazov — APPROVED** [review]
LGTM!

**Review by junliume — APPROVED** [review]

**Comment by qianfengz** [comment]
> ./bin/test_reduce_test --float --D 8 2 1 --R 0 1 2 --ReduceOp 0 --CompType 1 --N 0 --I 0 --scales 1 0



> @atamazov there are some test verification errors (_**verify_reduce_no_indices failed**_):
> 
> ```
> ./bin/test_reduce_test --float --D 8 2 1 --R 0 1 2 --ReduceOp 0 --CompType 1 --N 0 --I 0 --scales 1 0
> error: 0
> Max diff: 0
> verify_reduce_no_indices failed
> Input Tensor 8, 2, 1
> ```
> 
> at the same time the MIOpenDriver commands succeeded:
> 
> ```
> root@hpe-sjc2-43:/root/MIOpen/build# ./bin/MIOpenDriver reduce -A 1 -B 0 -C 1 -D 8,2,1 -I 0 -N 1 -O 0 -R 0,1,2
> MIOpenDriver reduce -A 1 -B 0 -C 1 -D 8,2,1 -I 0 -N 1 -O 0 -R 0,1,2
> ReduceTensor() Verifies on CPU and GPU (err=0.000000)
> ```

@jun  The ”verify_reduce_no_indices failed“ is printed because when the "verbose" mode is enable by "-v",   the common codes in test/driver.hpp  line 835 will call the fail() member function of specific test module.    This is a long-term issue,   one correct answer for the issue could be for each fail() member function to check the passed "int" value, and if the "int" value indicates only "verbose" or "timing", no failure information should be printed,  and the common verify() codes should always pass  a correct "int" value to the fail() interface.

**Comment by atamazov** [comment]
@junliume Let's push CI tests until completion. I'm not sure if test_reduce_custom will pass on all GPUs.

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -1546,6 +1546,29 @@ COMMAND ${CONV_CK_IGEMM_FWD_V6R1_DLOPS_NCHW_ENV}     $<TARGET_FILE:test_conv2d>
...
+COMMAND $<TARGET_FILE:test_reduce_test> ${MIOPEN_TEST_FLOAT_ARG} --scales 1 0 --CompType 1 --D 2048 32 1 --I 0 --N 1 ---ReduceOp 3 --R 0 2 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_reduce_test> ${MIOPEN_TEST_FLOAT_ARG} --scales 1 0 --CompType 1 --D 4 3 1 --I 0 --N 1 ---ReduceOp 2 --R 1 2 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_reduce_test> ${MIOPEN_TEST_FLOAT_ARG} --scales 1 0 --CompType 1 --D 3 4 1 --I 0 --N 1 ---ReduceOp 2 --R 0 2 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_reduce_test> ${MIOPEN_TEST_FLOAT_ARG} --scales 1 0 --CompType 1 --D 2048 32 1 --I 0 --N 1 ---ReduceOp 2 --R 0 2 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_reduce_test> ${MIOPEN_TEST_FLOAT_ARG} --scales 1 0 --CompType 1 --D 3 4 1 --I 0 --N 1 ---ReduceOp 2 --R 0 2 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_reduce_test> ${MIOPEN_TEST_FLOAT_ARG} --scales 1 0 --CompType 1 --D 12 11 1 --I 0 --N 1 ---ReduceOp 0 --R 0 1 2 ${MIOPEN_TEST_FLAGS_ARGS}
+COMMAND $<TARGET_FILE:test_reduce_test> ${MIOPEN_TEST_FLOAT_ARG} --scales 1 0 --CompType 1 --D 13 4 7 7 --I 0 --N 1 ---ReduceOp 0 --R 0 1 2 3 ${MIOPEN_TEST_FLAGS_ARGS}
```
Comment (atamazov) [inline test/CMakeLists.txt:1569]: `---ReduceOp` should be changed to `--ReduceOp`

**Pull request merged**